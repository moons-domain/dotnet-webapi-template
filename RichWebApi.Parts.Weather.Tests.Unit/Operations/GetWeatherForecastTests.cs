using MediatR;
using Microsoft.Extensions.DependencyInjection;
using RichWebApi.Entities;
using RichWebApi.Persistence;
using RichWebApi.Tests;
using RichWebApi.Tests.Logging;
using RichWebApi.Tests.Resources;
using Xunit.Abstractions;

namespace RichWebApi.Operations;

public class GetWeatherForecastTests : UnitTest
{
	private readonly IResourceScope _testResources;
	private readonly IServiceProvider _serviceProvider;

	public GetWeatherForecastTests(ITestOutputHelper testOutputHelper,
	                               ResourceRepositoryFixture resourceRepository,
	                               UnitDependencyContainerFixture container) : base(testOutputHelper)
	{
		_testResources = resourceRepository.CreateTestScope(this);
		var parts = new AppPartsCollection()
			.AddWeather();
		_serviceProvider = container
			.WithXunitLogging(TestOutputHelper)
			.WithTestScopeInMemoryDatabase(parts)
			.ConfigureServices(s => s.AddAppParts(parts))
			.BuildServiceProvider();
	}

	[Fact]
	public async Task LoadsFirstPage()
	{
		using var resources = _testResources.CreateMethodScope(); 
		var entities = resources.GetJsonInputResource<WeatherForecast[]>("entities");
		await _serviceProvider
			.GetRequiredService<IRichWebApiDatabase>()
			.PersistEntitiesAsync(entities);

		var input = resources.GetJsonInputResource<GetWeatherForecast>();
		var mediator = _serviceProvider.GetRequiredService<IMediator>();

		var result = await mediator.Send(input);
		resources.CompareWithJsonExpectation(TestOutputHelper, result);
	}

	[Fact]
	public async Task LoadsOne()
	{
		using var resources = _testResources.CreateMethodScope(); 
		var entities = resources.GetJsonInputResource<WeatherForecast[]>("entities");
		await _serviceProvider
			.GetRequiredService<IRichWebApiDatabase>()
			.PersistEntitiesAsync(entities);

		var input = resources.GetJsonInputResource<GetWeatherForecast>();
		var mediator = _serviceProvider.GetRequiredService<IMediator>();

		var result = await mediator.Send(input);
		resources.CompareWithJsonExpectation(TestOutputHelper, result);
	}
}