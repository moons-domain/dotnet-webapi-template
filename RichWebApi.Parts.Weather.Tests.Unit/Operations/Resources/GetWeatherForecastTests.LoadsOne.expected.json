{
  "Items": [
    {
      "Date": "2023-01-01T00:00:00",
      "TemperatureC": 14,
      "TemperatureF": 57,
      "Summary": "Summary"
    }
  ],
  "Page": 0,
  "Size": 1,
  "Total": 3
}